cpp/src/arrow/dataset/file_base.cc

```diff
@@ -243,20 +244,21 @@ std::string FileSystemDataset::ToString() const {
 
 void FileSystemDataset::SetupSubtreePruning() {
   subtrees_ = std::make_shared<FragmentSubtrees>();
-  SubtreeImpl impl;
+  compute::SubtreeImpl impl;
 
   auto encoded = impl.EncodeFragments(fragments_);
 
-  std::sort(encoded.begin(), encoded.end(),
-            [](const SubtreeImpl::Encoded& l, const SubtreeImpl::Encoded& r) {
-              const auto cmp = l.partition_expression.compare(r.partition_expression);
-              if (cmp != 0) {
-                return cmp < 0;
-              }
-              // Equal partition expressions; sort encodings with fragment indices after
-              // encodings without
-              return (l.fragment_index ? 1 : 0) < (r.fragment_index ? 1 : 0);
-            });
+  std::sort(
+      encoded.begin(), encoded.end(),
+      [](const compute::SubtreeImpl::Encoded& l, const compute::SubtreeImpl::Encoded& r) {
```

Comment:
Since this sort doesn't refer to anything in this scope it should move into SubtreeImpl. Otherwise other consumers of SubtreeImpl would need to reiterate the sort